Continue http cache work

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

Work in progress, and not quite worth a review yet. (Continuation of https://github.com/servo/servo/pull/4117)

TODO

- [ ] cache metadata (find some subset of`net_traits::Metadata` that can be shared across threads, it seems the problem is mainly stuff inside `hyper::header` in the `headers` field)

- [ ] determine which other fields of a `Response` need to be cached, so a full and valid one can be returned upon a cache hit. 

- [ ] determine how to best share the cache across fetch threads (inside HttpState like I tried now?)

- [ ] Spend more time reading the spec and make sure the cache follows it where it matters.

- [ ] Make the current wpt tests pass.

- [ ] More...

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #12972  (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18676)
<!-- Reviewable:end -->

jdm committed:
Beginnings of a http cache

Doom cache entries based on the initial response, and prevent matching against doomed cache enties.

Evict cache entries that have passed their expiry date instead of matching them.

Document the cache. Refactor incomplete entries to lessen Option-itis.

Revalidate expired cache entries instead of unconditionally evicting them.

Forbid missing docs in cache code.

Revalidate must-revalidate entries.

Fetch content tests from a local HTTP server.

Track requests made to the test HTTP server.

Add a simple test that a cached resource with no expiry is not revalidated. Correct inverted expiry check in revalidation code.

Fix incorrect revalidation logic that dropped the consumer channels on the floor.

Ensure that requests are cached based on their request headers.

Run a separate http server instance for each test to avoid intermittent failures due to concurrent cache tests.

Add a test for uncacheable responses.

Address review comments.

gterzian committed:
more http cache work

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/net/http_loader.rs, components/net/resource_thread.rs, components/net/lib.rs, components/net/http_cache.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify net code, but no tests are modified. Please consider adding a test!

gterzian commented:
Ok there is still a lot of work to be done, however I now have all the tests in `tests/wpt/web-platform-tests/fetch/http-cache/status.html` passing, so I guess it's a milestone of sorts. 

I would like to ask for initial feedback on one early design choice, relating to "how to plug the cache into fetch". Let me sketch the current design:

1. Fetch A starts for resource A. 
2. http_network_or_cache_fetch misses the cache, and starts a network fetch. 
3. The network fetch immediatly returns a `Response` A, with a body of type `ResponseBody::Receiving`.
4. Resonse A is cached. 
5. Fetch B starts for resource A.
6. Cache hit returns Response A, with a body still in `ResponseBody::Receiving`.
7. Fetch A receives the body of Response A from the network fetch thread, via `DoneChannel`.
8. Fetch A receives the full body and Response now has a body of type ResponseBody::Done.
9. Fetch A updates the cache with Response A, the cache only updates it's cached resource with the new body data. 
9. The cache, using the same `DoneChannel` that the network fetch would normally use, updates Fetch B by sending the whole body in one chunk. 
10. Fetch B now also has Response A with a full body, and updates the cache, not knowing whether the update via the channel came from the network or from the cache. The cache ignores this update(since the cached resource already has a body of type ResponseBody::Done).

I tried a fully sync cache operation first, however the tests were failing and I'm pretty sure it was because since the fetch is async, the cache would simply always miss since the full response hadn't been received yet by the time the second request testing for a cache hit came in. 
So I had to introduce the async element by returning a response immediately and updating the body using the channel. Re-using the network fetch channel seemed to make sense in the context of 'network_or_cache_fetch', essentially it's transparent from the main fetch code perspective.
The cache is only updated once the full body has been received, in the main fetch code(I'm not sure if doing it in chunks would be better). I tried updating from within the fetch thread, but it created some issues about sharing the request which isn't thread-safe and is necessary to update the cache for a given resource. 

How does the above look?

KiChjang commented on components/net/http_cache.rs:
```diff
@@ -187,19 +180,34 @@ fn response_is_cacheable(metadata: &Metadata) -> bool {
-                },
-                _ => (),
+        let has_no_cache_directive = directives.iter().any(|directive| {
+            if let header::CacheDirective::NoCache = *directive {
```
Why does `header::CacheDirective::NoCache == *directive` not work?

gterzian replied:
it does work, thanks!

gterzian commented:
@KiChjang you're right, thanks!

gterzian commented on components/net/http_loader.rs:
```diff
@@ -1366,14 +1383,20 @@ fn has_credentials(url: &ServoUrl) -> bool {
 }
 
 fn is_no_store_cache(headers: &Headers) -> bool {
+    println!("debug headers {:?}", headers);
```
@KiChjang In the context of the `http-cache/cc-request.html` tests, it seems that the request headers set in the tests don't come through to the headers in the request inside `network_or_cache_fetch`. It seems that some basic headers are set, but not the ones the test is adding. Could this be a bug in the tests, or is some part of the fetch code dealing with request headers not implemented yet?

```
 "debug headers Headers { Accept: */*"
", Accept-Language: en-US, en; q=0.5"
", Referer: http://web-platform.test:8000/fetch/http-cache/cc-request.html"
", User-Agent: Servo/wptrunner"
", }"
```

jdm replied:
https://dxr.mozilla.org/servo/rev/04f787dbf9b235bf3af58dc7f3160689cdb2311c/components/script/dom/request.rs#292-343 is the code that is supposed to propagate the headers specified in the test into the network stack.

gterzian replied:
thanks, I'll check it out...

gterzian replied:
fixed in https://github.com/servo/servo/pull/18676/commits/f4f7b9f54f0c854f468cd37eae967f26798789e6

The headers were not being copied from the dom Request to the net Request...

jdm commented:
I'm going to take more time on Monday, but I'll say that in general your description above looks like what I was expecting. For the purposes of merging the first implementation of the PR, I would like to have a plan for how we'll get to simultaneously updating other cache listeners while the original response is still ongoing, because it will be an important use case for the media stack that we're working towards integrating.

gterzian commented:
@jdm OK. 

Do you mean that the media stack would be listening on a channel other than the `DoneChannel` used by the fetch method? If so, perhaps we could store another vec of `Sender`, for the media stack, which would be updated in a [similar way to this](https://github.com/gterzian/servo/blob/929084c584ccf4513fe7dfb6f65a0a305d831131/components/net/http_cache.rs#L330). 

The fetch thread already does something similar by sending the response chunks over to the devtools in addition to the fetch method, see https://github.com/gterzian/servo/blob/929084c584ccf4513fe7dfb6f65a0a305d831131/components/net/http_loader.rs#L1107

I guess the integration would need to be done as well within the fetch thread, not just the cache? 

Current the devtool chan is stored on `FetchContext`, perhaps that could be the place to store a media stack channel?

jdm commented:
What I meant is that if a media element starts fetching a 100mb file, then seeks to some other point and causes a new request to retrieve the cached response, we don't want to have to wait until the original request is complete before we service the second request.

gterzian commented:
@jdm Thanks for the explanation. I'm wondering if this doesn't take us into the territory of [range requests](https://tools.ietf.org/html/rfc7233) and [caching partial content](https://tools.ietf.org/html/rfc7234#section-3.3)...

jdm commented:
It definitely does. My goal in this work is to ensure that we think about how those fit in with this design before merging it to ensure that we don't have to fundamentally change things later to support them.

KiChjang commented on components/net/http_cache.rs:
```diff
@@ -43,6 +43,12 @@ impl CacheKey {
             request_headers: request.headers
                                       .iter()
+                                      .filter(|header| {
+                                          match header.name().to_lowercase().as_ref() {
```
I don't think you need the `as_ref` call here if you have the `ref` keyword in the line above for `header`.

gterzian replied:
I tried that, and also putting a & in front of `header.name()`, and both would give 
```
expected type `&std::string::String`
              found type `&'static str`
```

gterzian commented:
Ok, regarding integration with the media stack:

1. The way the cache currently updates responses follows the model of the fetch worker thread(the difference being that for now the update is done in one chunk when the full response has been received, but it could be adapted to chunks as well), so it can grow along with it I guess.
2. If range requests fit the model of the media element(I still have to check out the spec), then we can implement support for that in the cache, there is a suite of tests covering those. 

Besides this, currently the entire http_cache test suite is passing, with the exception of the partial requests bits. I tried to clean up the code a little bit, still somewhat rough but ready for a first review I guess. 

I can work on the partial requests next weekend.

jdm commented:
Amazing!

gterzian commented:
I found a couple of subtle bugs, in fact a lot relied on the fact that the headers were part of the key, and while that gave us some cache logic "for free", it prevented some edge cases(for example not caching a response with `Vary: "*"`, for which I'll add tests later). 

Acache key that only contains the url,  actually follows the spec more closely:

```
   The primary cache key consists of the request method and target URI.
   However, since HTTP caches in common use today are typically limited
   to caching responses to GET, many caches simply decline other methods
   and use only the URI as the primary cache key.
```
[https://tools.ietf.org/html/rfc7234#section-4.1](https://tools.ietf.org/html/rfc7234#section-4.1)

I also realized that the entire body updating via the done channel was unecessary, in fact that code didn't work, and it was hidden by the fact that the headers in the key would prevent a match, so in fact the buggy code was just never called...

I do wonder how the response body is propagated without this, perhaps thanks to the call to `mem::update` in the fetch worker thread?

In any case, switched to using a ServoUrl as cachekey(not stripping fragments yet), and storing a Vec of resources per key.

 Some tests were only passing because the request headers were different. Also fixed the Vary calculation, again it didn't actually work, but that was hidden by the headers in they key that would prevent matches.

Almost all tests are passing again, except a few regarding invalidation, I think I just need to debug the function and adapt it to fit the latest...

gterzian commented:
Ok, all tests are passing again...

gterzian commented:
I'm puzzled as to why fetch::test_fetch_redirect_updates_method started hanging after https://github.com/servo/servo/pull/18676/commits/2185bf27172f4eaa7a20b28f7d119b3ee57a9821
seems there is some weird interaction between the cache, entry invalidation, and redirects. Fixed it for now, but not sure what the original problem was...

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+fn get_expiry_adjustment_from_request_headers(request: &Request, expires: Duration) -> Duration {
+    if let Some(directive_data) = request.headers.get_raw("cache-control") {
+        let directives_string = String::from_utf8(directive_data[0].to_vec()).unwrap();
+        let directives: Vec<&str> = directives_string.split(",").collect();
```
Lots of uses of unwrap/collect/to_vec in this function that we should avoid if possible.

gterzian replied:
Ok, I'll look into it...

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+
+/// Request Cache-Control Directives https://tools.ietf.org/html/rfc7234#section-5.2.1
+fn get_expiry_adjustment_from_request_headers(request: &Request, expires: Duration) -> Duration {
+    if let Some(directive_data) = request.headers.get_raw("cache-control") {
```
```rust
let directive_data = match request.headers.get_raw("cache-control") {
    Some(data) => data,
    None => return expires,
};
```

gterzian replied:
thanks

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+    // TODO: since this cache is shared, taking proxy-revalidate into account
+    // https://tools.ietf.org/html/rfc7234#section-5.2.2.7
+    let has_expired = (adjusted_expires < time_since_validated)
+        | (adjusted_expires == time_since_validated);
```
Let's use ||, and move the operator to the previous line.

gterzian replied:
Ok

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+    response.body = cached_resource.body.clone();
+    response.status = cached_resource.status.clone();
+    response.raw_status = cached_resource.raw_status.clone();
+    response.url_list = cached_resource.url_list.clone();
```
We'll want to copy:
* the https state of the cached response
* the referrer and referrer policy of the request into the response

gterzian replied:
OK

gterzian replied:
done, see lines below.

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+    status: Option<StatusCode>,
+    raw_status: Option<(u16, Vec<u8>)>,
+    url_list: Vec<ServoUrl>,
+    expires: Arc<Mutex<Duration>>,
```
Why does this field need its own mutex?

gterzian replied:
You're right that's a mistake, I think I did as a kneejerk reaction when being told by the compiler that the field wasn't mutable, instead of assigning it to a mutable let.

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+    /// Character set.
+    pub charset: Option<String>,
+    /// Headers
+    pub headers: Arc<Mutex<Headers>>,
```
Why do the headers need their own mutex?

gterzian replied:
I did that because some internal part of hyper::Headers didn't implement the right traits to make it shareable across threads.

KiChjang replied:
Why do you need to share the headers across threads? Aren't you supposed to be cloning them instead?

gterzian replied:
The problem seems to occur here(and a few other places where a thread is build) https://github.com/gterzian/servo/blob/ab5c9798a03cdcf24704fd418cf9db827e3d3514/components/net/resource_thread.rs#L337

```
error[E0277]: the trait bound `std::cell::UnsafeCell<serde::export::Option<std::vec::Vec<std::vec::Vec<u8>>>>: std::marker::Sync` is not satisfied in `(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)`
   --> /Users/gregory/Projects/servo/components/net/resource_thread.rs:337:83
    |
337 |         thread::Builder::new().name(format!("fetch thread for {}", req_init.url)).spawn(move || {
    |                                                                                   ^^^^^ `std::cell::UnsafeCell<serde::export::Option<std::vec::Vec<std::vec::Vec<u8>>>>` cannot be shared between threads safely
    |
    = help: within `(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)`, the trait `std::marker::Sync` is not implemented for `std::cell::UnsafeCell<serde::export::Option<std::vec::Vec<std::vec::Vec<u8>>>>`
    = note: required because it appears within the type `hyper::header::internals::cell::OptCell<std::vec::Vec<std::vec::Vec<u8>>>`
    = note: required because it appears within the type `hyper::header::internals::item::Item`
    = note: required because it appears within the type `(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)`
    = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)>`
    = note: required because it appears within the type `alloc::raw_vec::RawVec<(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)>`
    = note: required because it appears within the type `std::vec::Vec<(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)>`
    = note: required because it appears within the type `hyper::header::internals::vec_map::VecMap<unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item>`
    = note: required because it appears within the type `hyper::header::Headers`
    = note: required because it appears within the type `http_cache::CachedMetadata`
    = note: required because it appears within the type `http_cache::CachedResource`
    = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<http_cache::CachedResource>`
    = note: required because it appears within the type `alloc::raw_vec::RawVec<http_cache::CachedResource>`
    = note: required because it appears within the type `std::vec::Vec<http_cache::CachedResource>`
    = note: required because of the requirements on the impl of `std::marker::Sync` for `std::collections::hash::table::RawTable<http_cache::CacheKey, std::vec::Vec<http_cache::CachedResource>>`
    = note: required because it appears within the type `std::collections::HashMap<http_cache::CacheKey, std::vec::Vec<http_cache::CachedResource>>`
    = note: required because it appears within the type `http_cache::HttpCache`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::RwLock<http_cache::HttpCache>`
    = note: required because it appears within the type `http_loader::HttpState`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<http_loader::HttpState>`
    = note: required because it appears within the type `[closure@/Users/gregory/Projects/servo/components/net/resource_thread.rs:337:89: 363:10 req_init:net_traits::request::RequestInit, http_state:std::sync::Arc<http_loader::HttpState>, ua:std::borrow::Cow<'_, str>, dc:serde::export::Option<std::sync::mpsc::Sender<devtools_traits::DevtoolsControlMsg>>, filemanager:filemanager_thread::FileManager, res_init_:serde::export::Option<net_traits::response::ResponseInit>, sender:ipc_channel::ipc::IpcSender<net_traits::FetchResponseMsg>]`
```

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+/// The key used to differentiate requests in the cache.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct CacheKey {
+    url: ServoUrl
```
I saw that you originally included headers in the key, then later removed them. What was the issue, exactly? Could we just omit the vary-related headers when comparing them and checking the key?

gterzian replied:
One problem was that I ended up doing a lot of looping through all entries and comparing `key.url()` with `request.url()`, for example in the context of vary, since the headers will change from request to request, while the url stays the same, and we still need to match those.

The problem with excluding vary headers is that on the request side you don't know which headers it will be(it is only the response that contains a `Vary` header, containing the name of the relevant request headers). Also, if the second requests contains additional headers, unrelated to the Vary check, you need to ignore those. If the headers are in the key, such request will be excluded, hence I was looping and comparing url.

Same issue with invalidation, one needs to invalidate based on the url(or value of `Location` or `Content-Location` header, which are urls as well).

I did also run into issues related to request `cache-control` headers, which I did initially successfully exclude when building they key. 

So based on those use cases, it felt more natural to have urls as keys. The spec also points in that direction.

jdm reviewed (changes requested):
This feels like it's going in the right direction! I have a couple questions about some of the design choices, so I'll probably wait to provide more feedback until I understand those.

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+    }
+
+    fn from_url_string(url: String) -> CacheKey {
+        let url = Url::parse(&url).unwrap();
```
Ok

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+
+
+/// Determine if a status code corresponds to a response for a redirected request.
+fn is_redirect(status: StatusCode) -> bool {
```
yep, I'll make the other one public and import it.

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+fn calculate_response_age(response: &Response) -> Duration {
+    // TODO: follow the spec more closely (Date headers, request/response lag, ...)
+    if let Some(secs) = response.headers.get_raw("Age") {
+        let seconds = String::from_utf8(secs[0].to_vec()).unwrap();
```
thanks

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+    // TODO: follow the spec more closely (Date headers, request/response lag, ...)
+    if let Some(secs) = response.headers.get_raw("Age") {
+        let seconds = String::from_utf8(secs[0].to_vec()).unwrap();
+        return Duration::seconds(seconds.parse::<i64>().unwrap());
```
OK

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+
+    /// https://tools.ietf.org/html/rfc7234#section-4 Constructing Responses from Caches.
+    pub fn construct_response(&self, request: &Request)
+        -> Option<CachedResponse> {
```
ok

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+    pub fn construct_response(&self, request: &Request)
+        -> Option<CachedResponse> {
+        // TODO: generate warning headers as appropriate https://tools.ietf.org/html/rfc7234#section-5.5
+        match request.method {
```
yep

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+        if let Some(ref resources) = self.entries.get(&entry_key) {
+            for cached_resource in resources.iter() {
+                let mut can_be_constructed = true;
+                if let Ok(ref mut stored_headers) = cached_resource.metadata.headers.try_lock() {
```
No reason, I had a deadlock at some point in another part so ended up writing code like that, but I don't think it's necessary here.

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+                                    // If a header field is absent from a request,
+                                    // it can only match another request if it is also absent there.
+                                    can_be_constructed =
+                                        !cached_resource.request_headers.iter().any(|&(ref name, _)| {
```
thanks

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+                filtered: FilteredMetadata::Cors(metadata),
+                unsafe_: _ }) |
+            Ok(FetchMetadata::Unfiltered(metadata)) => {
+                if response_is_cacheable(&metadata) {
```
awesome

gterzian commented on components/net/http_loader.rs:
```diff
@@ -897,64 +900,86 @@ fn http_network_or_cache_fetch(request: &mut Request,
+        let complete_http_response_from_cache = http_cache.construct_response(&http_request);
+        if http_request.cache_mode != CacheMode::NoStore &&
+           http_request.cache_mode != CacheMode::Reload {
+            // TODO Substep 1 and 2. Select a response from HTTP cache.
```
yep

gterzian commented on components/net/http_loader.rs:
```diff
@@ -897,64 +900,86 @@ fn http_network_or_cache_fetch(request: &mut Request,
-        if http_request.cache_mode == CacheMode::OnlyIfCached {
-            return Response::network_error(
-                NetworkError::Internal("Couldn't find response in cache".into()))
+        // Note: do we need to set http_request.cache_mode to CacheMode::OnlyIfCached here or elsewhere?
```
Ok

gterzian commented on components/net/http_loader.rs:
```diff
@@ -897,64 +900,86 @@ fn http_network_or_cache_fetch(request: &mut Request,
+            response_504.status = Some(StatusCode::GatewayTimeout);
+            response_504.raw_status = Some((504, b"Gateway Timeout".to_vec()));
+            response_504.body = Arc::new(Mutex::new(ResponseBody::Empty));
+            response = Some(response_504);
```
The network error wouldn't result in a 504 response being propagated to the javascript code, and the relevant test was failing. But perhaps this isn't the right way to fix it.

jdm replied:
Yeah, that sounds like an incorrect test to me. Which file is it?

gterzian replied:
https://github.com/servo/servo/blob/f0b8dad063b459db12ac27d44afff32cd6923d30/tests/wpt/web-platform-tests/fetch/http-cache/cc-request.html#L198

jdm replied:
I have filed https://github.com/w3c/web-platform-tests/issues/8032.

gterzian commented on components/net/http_loader.rs:
```diff
@@ -897,64 +900,86 @@ fn http_network_or_cache_fetch(request: &mut Request,
-                NetworkError::Internal("Couldn't find response in cache".into()))
+        // Note: do we need to set http_request.cache_mode to CacheMode::OnlyIfCached here or elsewhere?
+        let mut only_if_cached = false;
+        if let Some(directive_data) = http_request.headers.get_raw("cache-control") {
```
CacheMode::OnlyIfCached was always false. So I replaced that check  by cecking the request headers myself.

jdm replied:
I do not believe they are related concepts.

gterzian replied:
Ok, removed this change.

gterzian commented on components/script/dom/request.rs:
```diff
@@ -475,6 +478,7 @@ fn is_method(m: &ByteString) -> bool {
         Some("connect") => true,
         Some("options") => true,
         Some("trace") => true,
+        Some(_) => true,
```
Ok, I did this because some wpt cache tests require handling of unknown request methods. So I'll remove and we can fix this separately.

gterzian replied:
Ok, I did this because some wpt cache tests require handling of unknown request methods. So I'll remove and we can fix this separately.

gterzian replied:
example of a test that would fail https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/fetch/http-cache/invalidate.html#L71

jdm replied:
I meant that for the purposes of this PR we can replace the definition of `is_method` with `m.as_str().is_some()`.

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+        let entry_key = CacheKey::new(request.clone());
+        if let Some(cached_resources) = self.entries.get_mut(&entry_key) {
+            for cached_resource in cached_resources.iter_mut() {
+                if let Ok(ref mut stored_headers) = cached_resource.metadata.headers.try_lock() {
```
nope. no reason I think.

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+        if let Some(&header::Location(ref location)) = response.headers.get::<header::Location>() {
+            self.invalidate_for_url(location.clone());
+        }
+        // TODO: update hyper to use typed getter.
```
Ok, on this one they actually have done it already in the more recent versions. However the Age header still hasn't been done it seems, perhaps it can be implemented as part of the cache, since it only relates to cache operations I believe.

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+        None
+    }
+
+    fn invalidate_for_url(&mut self, url: String) {
```
yes I think so, in fact I would have just used cache keys directly except that in the context of `Location` and `Content-Location` headers containing the url to be invalidated, you can't use the request to build the key.

jdm replied:
What about `Location: foo.html`?

gterzian replied:
I'm under the impression that such a header value should be ignored according to the spec:

```

   However, a cache MUST NOT invalidate a URI from a Location or
   Content-Location response header field if the host part of that URI
   differs from the host part in the effective request URI (Section 5.5
   of [RFC7230]).  This helps prevent denial-of-service attacks.

```

jdm replied:
I'm not sure I understand how that text means that `Location: foo.html` is ignored. Location headers are parsed relative to the request's URL.

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+        if let Some(status) = response.status {
+            // Ignoring redirects.
+            // Note: without this, the fetch::test_fetch_redirect_updates_method would hang.
+            // An unknown interaction between the cache and redirects caused this problem.
```
Ok, I think this was actually a deadlock inside `construct_response`, since `create_cached_response` was locking the headers that were locked just before inside `construct_response`. I'm not sure why this only was apparent inside the revalidation of a re-direct though. 

Fix https://github.com/servo/servo/pull/18676/commits/1cef9cec53934d031b34c378461b25bc5ed79a63

Also removed the check for redirect status inside `invalidate`, unit test passing again without the check. https://github.com/servo/servo/pull/18676/commits/bd10d895dc19052af4940177d34ef89eb09ac655

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,488 @@
+                    let request_headers = request.headers
+                        .iter()
+                        .map(|header| (String::from_str(header.name()).unwrap_or(String::from("None")),
+                                                                  header.value_string()))
```
I ended up storing the request headers in their original `hyper::Headers` form, like those of the response. It simplifies the vary code as well...

gterzian commented:
Thanks for this first review, everything has been addressed(including that weird bug that made the redirect unit test hanging).

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18924) made this pull request unmergeable. Please resolve the merge conflicts.

gterzian commented:
I'm planning on adding support for partials this weekend, did you have a chance to look at the changes I made in response to the reviews so far? 

Question in terms of squashing: should I keep commits that modify wpt tests separate?

Currently the build is failing because of the conflict I think, I prefer not to rebase in the middle of a review as doing so marks all comments as 'outdated'...

jdm commented:
I'm going to review it all again today. Commits that modify tests do not need to be kept separate if the make sense to squash with others.

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,486 @@
+    response.url_list = cached_resource.url_list.clone();
+    response.https_state = cached_resource.https_state.clone();
+    response.referrer = cached_resource.referrer.clone();
+    response.referrer_policy = cached_resource.referrer_policy.clone();
```
This doesn't look right - the referrer and referrer policy can vary in the request that's being made; we shouldn't be caching them and reusing them later.

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,486 @@
+                            // If a header field is absent from a request,
+                            // it can only match a stored response if those headers,
+                            // were also absent in the original request.
+                            can_be_constructed = !original_request_headers.get_raw(vary_val).is_some();
```
Instead of !a.is_some(), prefer a.is_none().

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,486 @@
+
+    fn from_url_string(url: &str) -> Result<CacheKey, ()> {
+        if let Ok(url) = Url::parse(&url) {
+            if !url.has_host() {
```
If I understand correctly, this is supposed to address my question about `Location: foo.html`? As my response (hopefully) explained, I don't believe this is the correct solution; this code would not be reached, because Url::parse would return an error for a relative URL.

gterzian replied:
Ok I see, would it make sense if I were to use the concat the request origin with the value of the location header as a way to find the relevant entry by url?

jdm replied:
Use `request.current_url().join(url)`; that uses the request's current URL as the base and parses the string relative to it. A non-relative URL will override it, too.

gterzian replied:
thanks

jdm commented on components/net/http_loader.rs:
```diff
@@ -897,34 +900,38 @@ fn http_network_or_cache_fetch(request: &mut Request,
-            // response = http_request
-        }
+    if let Ok(http_cache) = context.state.http_cache.read() {
+        let complete_http_response_from_cache = http_cache.construct_response(&http_request);
```
Add `// Substeps 1 and 2` above this line.

jdm commented on tests/wpt/web-platform-tests/fetch/http-cache/heuristic.html:
```diff
@@ -35,8 +36,7 @@
             response_headers: [
-              ['Last-Modified', http_date(-3 * 100)],
-              ['Cache-Control', 'public']
+              ['Last-Modified', http_date(-3 * 100)]
```
Why is this change required?

gterzian replied:
The test reads "when Cache-Control: public is not present", and leaving the "public" header in the response makes the test always fail(since the response will be cached because of it).

jdm commented on tests/wpt/web-platform-tests/fetch/http-cache/vary.html:
```diff
@@ -259,6 +261,57 @@
+        ]
+      },
+      {
+        name: "HTTP cache doesn't use Vary response a field value of '*'.",
```
`with a field`

jdm commented on tests/wpt/web-platform-tests/fetch/http-cache/vary.html:
```diff
@@ -245,11 +246,12 @@
       },
       {
-        name: "HTTP cache doesn't use three-way Vary response when request omits variant header.",
+        name: "HTTP cache doesn't use three-way Vary response when request doesn't match, regardless of header order.",
```
Are we losing coverage for the original test by changing it like this?

gterzian replied:
I don't think so, I wrote the test when I realized my code was relying on the order of headers, and it failed while the other would still pass.

gterzian replied:
by the way, I left the original test in there.

jdm commented on components/net/http_loader.rs:
```diff
@@ -828,7 +832,6 @@ fn http_network_or_cache_fetch(request: &mut Request,
             if !http_request.headers.has::<Pragma>() {
                 http_request.headers.set(Pragma::NoCache);
             }
-
```
nit: revert this change.

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,486 @@
+
+#![deny(missing_docs)]
+
+//! A memory cache Implements the logic specified in http://tools.ietf.org/html/rfc7234
```
`A memory cache that implements the logic specified in <http://tools.ietf.org/html/rfc7234>`

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,486 @@
+        return false;
+    };
+    let headers = metadata.headers.as_ref().unwrap();
+    if let Some(_) = headers.get::<header::Expires>() {
```
nit: `if headers.has::<header::Expires>() {`

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,486 @@
+        is_cacheable = true;
+    }
+    match headers.get::<header::CacheControl>() {
+        Some(&header::CacheControl(ref directive)) => {
```
nit: `if let Some(&header::CacheControl(ref directive)) = headers.get::<header::CacheControl>() {`

jdm commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,486 @@
+        None => ()
+    }
+    match headers.get::<header::Pragma>() {
+        Some(&header::Pragma::NoCache) => {
```
nit: if let

jdm commented on components/net/http_cache.rs:
```diff
@@ -237,7 +237,7 @@ fn get_expiry_adjustment_from_request_headers(request: &Request, expires: Durati
     for directive in directives_string.split(",") {
         let mut directive_info = directive.split("=");
-        match (directive_info.nth(0), directive_info.nth(1)) {
+        match (directive_info.nth(0), directive_info.nth(0)) {
```
It took me a while to figure out why this is correct - let's do something more explicit like `match (directive_info.next(), directive_info.next()) {`

jdm reviewed (changes requested):
This is looking great!

gterzian commented:
Ok I have addresssed the last points, and also made a first pass at adding support for range requests(the test are passing, even though I had to change them quite a bit, I think those hadn't been actually used in the past, or I got the spec wrong and `-5` doesn't mean 'the last five bytes').

jdm commented on components/net/http_cache.rs:
```diff
@@ -367,9 +368,165 @@ impl HttpCache {
+                        if let Some(&header::ContentRange(
+                            header::ContentRangeSpec::Bytes {
+                            range: Some((res_beginning, res_end)),
+                            instance_length: Some(total) })) = headers.get::<header::ContentRange>() {
```
Same match/continue transformation as previous comment.

jdm commented on tests/wpt/web-platform-tests/fetch/http-cache/partial.html:
```diff
@@ -57,11 +57,97 @@
+            ],
+            expected_type: "cached",
+            expected_status: 206,
+            expected_response_text: "90"
```
Why is this two bytes?

jdm reviewed (changes requested):
Range support is looking really great!

gterzian commented on tests/wpt/web-platform-tests/fetch/http-cache/partial.html:
```diff
@@ -51,9 +52,10 @@
           {
             request_headers: [
-              ['Range', "bytes=-1"]
+              ['Range', "bytes=0-1"]
```
The test was expecting a response text of "01", while using "-1" as range, which as I understand means 'the last byte'. So I could either change the expected text to "0", or change the range to "0-1"(which means the first and the second bytes, right?). By the way there is still a test with the "-1" a little bit below(I've also changed the last byte to make sure no one is fooled by the first and the last bytes in the response body matching).

jdm commented on components/net/http_cache.rs:
```diff
@@ -302,6 +297,185 @@ fn create_cached_response(request: &Request, cached_resource: &CachedResource, c
+        }
+    } else {
+        // Not a Range request.
+        if let Some(ref cached_resource) = candidates.first() {
```
I actually meant that the ranged check and non-ranged behavior could stay in the caller, just yo reduce the complexity of this function.

jdm reviewed (changes requested):
One little change and then I guess it's fine to rebase!

gterzian committed:
copy headers from script request to net request

gterzian commented:
Ok, rebased, and good for a test ride...

jdm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit e10b48dfeecbc3795ededf3e6443268179500176 with merge f6400d7e5dedc11991d07a25e6160068d5d1d9c8...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2584)

jdm commented:
See http://build.servo.org/grid (or http://build.servo.org/grid?width=20 if it's out of sight) and the column for f6400d7e5dedc11991d07a25e6160068d5d1d9c8 for the full set of failures on all builders.

gterzian commented:
Ok, lots of unpexcted PASS, but also some FAIL, will look into them...

gterzian commented:
still a work in progress, but so far:
1. It seems that https://github.com/servo/servo/pull/18676/commits/0c18c78883ae4bf404491fded0c0d426ac7fa16f fixed a bunch of cors related tests, so I've added the updated test expectations to it. Since cors, I think, relies on request headers, it seems that this fix indeed makes sense.
2. There were some cors, cookies, and referrer tests that started failing. I could fix them by adding a random number to their get params. I think it makes sense, and that the problem isn't the cache behavior, because those tests relied on the server setting some custom response headers based on the get params of the request, and the request headers, and the tests wer re-using a given url for different tests, so after the first one it started getting the the cached response, as opposed to the "magical" server response.

There are still some issues with authentication tests, actually they are passing now, and there are also some tests related to request cache mode which started failing(since they don't expect a cached response). I think those last ones actually related to what we came across previously were the CacheMode of the request wasn't set properly(and I started looking at the headers direction which you said was actually unrelated).

jdm commented:
I'm going to set aside some time to think about the cache key situation again.

gterzian committed:
pass cache_mode from request to request init

gterzian commented:
Ok, looking forward to your thoughts on the cache key.

In the meantime, I think all tests should be passing now, further changes required were:

1. Passing the CacheMode around from script to fetch(it was always Default before), see https://github.com/servo/servo/pull/18676/commits/008479a641102166c8989a803122307bb4ad9171.
2. Not storing a response if the request contains the 'no-store' directive, see https://github.com/servo/servo/pull/18676/files#diff-e980a3e3bf7e4cb8a8d3dda60ce7a408R629.

gterzian committed:
allow unknown HTTP methods

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit fd8b8c80d5c0ddc2639674f3c32549fb67db5728 with merge 00872de0288be6212f709343eb4031043b8e4269...

gterzian commented:
A small bonus: allowing for unknown methods like we discussed earlier(it got lost along the way), and more tests passing. https://github.com/servo/servo/pull/18676/commits/3a94d465e69f322cf131fe71dbad402310135ffe

gterzian commented:
seems like there a lot of `expected \"timeout\" but got \"load\""` in the XMLHttpRequest tests, which seems to be because the test expects a timeout, but gets a cached response immediately. I could actually solve it again by adding random GET params to the requests and preventing caches, however I had another look at the spec and I think I'm going to try to make caching dependent on either a "Expires" or "Last-Modified" header and see if it solves the problem(whithout breaking the http-cache test suite)

> 
   Note that, in normal operation, some caches will not store a response
   that has neither a cache validator nor an explicit expiration time,
   as such responses are not usually useful to store.  However, caches
   are not prohibited from storing such responses.

gterzian commented:
Ok, I was able to remove all those cache busters in the tests, simply by removing the branch that would cache a response only because it's status code was "cacheable by default" in https://github.com/servo/servo/pull/18676/files#diff-e980a3e3bf7e4cb8a8d3dda60ce7a408R100

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit bef9517c7f60d6d737e60cf13d4736fa935b1385 with merge fa8ee23d2189abac2c83d4fd226ffe0ac140a20f...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2562)

jdm commented:
I have a suspicion that some of the new passes in cors/ are intermittent passes.

gterzian commented on components/net/http_cache.rs:
```diff
@@ -0,0 +1,660 @@
+            let directives_string = String::from_utf8_lossy(&data[0]);
+            for directive in directives_string.split(",") {
+                let mut directive_info = directive.split("=");
+                if directive_info.next() == Some("no-cache") {
```
I just realized what I am doing here isn't correct, because although the 'no-cache' directive is set by fetch for a Reload or Nostore cache mode(and so this makes the relevant tests pass), the directive itself is unrelated to cache storage, so instead I am looking at the cachemode logic in the fetch code, I think it needs to be slightly adapted so that the cache is used or not based on the cachemode as described in https://fetch.spec.whatwg.org/#concept-request-cache-mode

gterzian commented:
Those CORS tests seem to be reliably passing on my machine too, I just didn't run and update them before...

gterzian commented:
Ah those Cors test are actually specifically testing cache behavior, looks like a genuine new pass.

gterzian commented:
Ok I think we can give it another try, I'll still have to squash those two last commits.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 1cd2e5f95a97c31aa846ca236ce31812395a1f41 with merge be675d45d843c8775cf805e025503debdd80b4bb...

jdm commented:
@bors-servo: retry

bors-servo commented:
:hourglass: Trying commit 1cd2e5f95a97c31aa846ca236ce31812395a1f41 with merge 78863012c0e449d3b317cafd9a1b87f331cecb8a...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/9783), [arm32](http://build.servo.org/builders/arm32/builds/10017), [arm64](http://build.servo.org/builders/arm64/builds/10047), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10065), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/6895), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/6921), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10123), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/2557), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2573), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/6818), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/6883), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2636), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/2598), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5440)
State: approved= try=True

gterzian commented:
Ok, squashed those last commits into https://github.com/servo/servo/pull/18676/commits/00879a42e63c52db56e542628f3b13bd8763855e

gterzian commented:
I was able to delete a few more FAIL expectations in the request cache tests, however it's weird because they weren't picked up by the test runner(it wouldn't show unexpected test results). I squashed the commit, but you can see the original at https://github.com/gterzian/servo/commit/89ba7286a1da78e405caf2aa3725b2dacbcd5994

gterzian commented:
Ok I was able to get those last only-if-cached tests to pass as well, key was only using that cachemode when request mode is same origin... It did require moving things around a bit, squashed it already but here is the diff(I also added a check for Etag to see if a response is worth caching): https://github.com/gterzian/servo/commit/761fb76c0795bca9de1e7e3f1d17a68f148dd52a

jdm commented:
Ok, last request and then I think this is good to merge - can we add a preference that controls whether the cache is enabled? That will make it much easier to measure performance benefits (or problems) from using it.

gterzian committed:
also cache redirects

gterzian commented:
Ok, two more commits to review(and to be squashed):

1. Also cache re-directs. This required using the current_url as cache key, and caching the location_url. 
2. Add a disbable-http-cache option. Actually I was not sure how to test this, how could one pass this option to the wpt runner?

By the way I briefly looked at it and we could also later easily disable/enable and clear the cache while servo is running, by having the ResourceChannelManager call methods on the cache in response to CoreResourceMsg. Maybe something similar could also be used to ask for and measure memory usage?

jdm commented on components/config/opts.rs:
```diff
@@ -143,6 +143,9 @@ pub struct Opts {
     pub sandbox: bool,
 
+    /// Whether to disable the http cache.
+    pub disable_http_cache: bool,
```
Rather than a command-line flag, we should use runtime preferences that are controlled by [prefs.json](https://github.com/servo/servo/blob/master/resources/prefs.json) and accessed by the [preferences api](https://github.com/servo/servo/blob/4cf2ce66fc4f970a47ab1fb4b9aa1a55282640f7/components/script/dom/mouseevent.rs#L176). This also allows us to write a WPT test that turns off the cache if we want: https://github.com/servo/servo/blob/234951e4b6d66891cd0cc9705f42e3ca16d9e3d5/tests/wpt/mozilla/meta/mozilla/mozbrowser/__dir__.ini#L1 . If we do that, it should live in tests/wpt/mozilla/tests/mozilla.

gterzian committed:
add pref to disable http cache, mozilla tests

gterzian commented:
Thanks for the pointers, ready for review.

jdm commented:
@bors-servo r+

bors-servo commented:
:clipboard: Looks like this PR is still in progress, ignoring approval

gterzian commented:
I think it needs re-approval after the title change...

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit dcc3adf has been approved by `jdm`

<!-- @bors-servo r=jdm dcc3adf953d30516243a54afbd760f9897c6d189 -->

bors-servo commented:
:hourglass: Testing commit dcc3adf953d30516243a54afbd760f9897c6d189 with merge 1aa5415f262b6351035510fd5f34d153ff34a229...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/6938)

KiChjang commented:
Uh oh:
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/http-cache.html:
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 200 response.
  │   → assert_equals: Response status expected 200 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 203 response.
  │   → assert_equals: Response status expected 203 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 204 response.
  │   → assert_equals: Response status expected 204 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 299 response.
  │   → assert_equals: Response status expected 299 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 400 response.
  │   → assert_equals: Response status expected 400 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 410 response.
  │   → assert_equals: Response status expected 410 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 499 response.
  │   → assert_equals: Response status expected 499 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 500 response.
  │   → assert_equals: Response status expected 500 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 502 response.
  │   → assert_equals: Response status expected 502 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 503 response.
  │   → assert_equals: Response status expected 503 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 504 response.
  │   → assert_equals: Response status expected 504 but got 404
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 599 response.
  │   → assert_equals: Response status expected 599 but got 404
  │ 
  └ make_test/</<.code/<@http://web-platform.test:8000/fetch/http-cache/http-cache.js:161:19

  ▶ Unexpected subtest result in /_mozilla/mozilla/http-cache.html:
  │ FAIL [expected PASS] HTTP cache, when disabled, does not cache a 404 response.
  │   → assert_equals: Response body expected "NTk0MjIzOTMzNjI1LjU5NjY=" but got "{\"error\": {\"message\": \"\", \"code\": 404}}"
  │ 
  └ make_test/</<.code/<@http://web-platform.test:8000/fetch/http-cache/http-cache.js:182:17
```

gterzian commented:
I had to copy `http-cache.js` over from the wpt folder to the mozilla one, and change the url to correctly point to the server. 
Should I perhaps also copy the server `http-cache.py` across as well so those two are isolated? 
Any licence/attribution necessary when copying across like that?

jdm commented:
I think this is fine as it is.
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 6196a6e has been approved by `jdm`

<!-- @bors-servo r=jdm 6196a6e65d03e7e4f9ff857f0c2315413a41e44a -->

bors-servo commented:
:hourglass: Testing commit 6196a6e65d03e7e4f9ff857f0c2315413a41e44a with merge e2bc0f017cfd24734777f64d8607b23dbe9552d7...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/9848), [arm32](http://build.servo.org/builders/arm32/builds/10082), [arm64](http://build.servo.org/builders/arm64/builds/10112), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10130), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/6961), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/6987), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10188), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/2622), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2638), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/6883), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/6948), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2702), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/2662), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5505)
Approved by: jdm
Pushing e2bc0f017cfd24734777f64d8607b23dbe9552d7 to master...

gterzian commented:
thanks for your help @jdm and @KiChjang

jdm commented:
Thanks for tackling this work!

Manishearth commented:
Wooo!

I added the `aborted` flag on Response, btw. Would you be able to do a followup to check that in the cache (or tell me where approximately it needs to be added)? Thanks!

gterzian commented:
@Manishearth great,  I can open a new PR to check for the `aborted` flag...
